### PR TITLE
[FSDP] Fix `keep_low_precision_grads=True` for `use_orig_params=True`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90028
* __->__ #90027

For any `flat_param.data = flat_param.to(...)` or `flat_param.grad.data = flat_param.grad.to(...)`, we must also refresh sharded parameter/gradient views, respectively, if the storage changes.

For `keep_low_precision_grads=True` and a sharded strategy, we cast the gradient back to the low precision using `.data` to bypass the PyTorch check that a parameter and its gradient have the same dtype. For `use_orig_params=True` before this PR, the gradient would incorrectly still be in full precision, not low precision, since we did not refresh views (this can actually be considered a memory leak since we have two copies of the gradient now, one in low precision and one in full precision). This PR refreshes the views.

### CODE REVIEW

File: torch/distributed/fsdp/flat_param.py [hunk 1]
```diff
@@ -1092,11 +1092,12 @@ def prepare_gradient_for_optim(self):
...
-                # dtype.
-                flat_param.grad.data = flat_param.grad.to(
-                    self._config.low_prec_param_dtype
-                )
+                if flat_param.grad.dtype != self._config.low_prec_param_dtype:
+                    flat_param.grad.data = flat_param.grad.to(
+                        self._config.low_prec_param_dtype
+                    )
+                    if self._use_orig_params:
+                        self._use_sharded_grad_views()
```
Comment (mrshenli) [inline torch/distributed/fsdp/flat_param.py:1100]: this looks like `gradient_as_bucket_view` in DDP. Would I be correct that when using low precision, grads are forced to be views. And when not using low precision, gradients can be standalone tensors?

File: torch/distributed/fsdp/flat_param.py [hunk 2]
```diff
@@ -1511,9 +1512,20 @@ def _use_sharded_grad_views(self) -> None:
...
-                        param.shape
-                    )
+                    if self._keep_low_precision_grads:
+                        # NOTE: This is a hack using `.data` to side step the
+                        # check that parameter/gradient dtypes match. Here,
+                        # `param` has full precision; `grad` has low precision.
+                        if param.grad is None:
+                            # `.grad` must have the same shape as `param`
+                            param.grad = torch.empty_like(param)
+                        param.grad.data = grad[
```
Comment (mrshenli) [inline torch/distributed/fsdp/flat_param.py:1522]: does this mean, after this assignment, param and param.grad are different dtypes?

**Review by mrshenli — APPROVED** [review]

File: torch/distributed/fsdp/flat_param.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:1522]: Yup.

File: torch/distributed/fsdp/flat_param.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:1100]: The gradients are always views regardless of the gradient dtype, but for `keep_low_precision_grads=True`, we have the hack to cast the gradient back to low precision from full precision at the end of the backward pass. This cast changes the flat gradient's storage, so we have to recreate views to point into the new storage here.

The typical code path is that the gradient views are created at the end of the post-backward hook, and that is all. This code snippet here is an extra re-creation of gradient views.

**Pull request closed without merging**